File: src/main/java/com/github/aivancioglo/resttest/http/OAuth2.java
package com.github.aivancioglo.resttest.http;

import com.github.aivancioglo.resttest.setters.OAuth1Token;

public class OAuth2 {
    private String token;

    /**
     * Use this method to get token.
     *
     * @return token.
     */
    public String token() {
        return token;
    }

    /**
     * Use this method to set/update token.
     *
     * @param token is sting value.
     */
    public void setToken(String token) {
        this.token = token;
    }

    /**
     * Add token to request.
     * @param token oAuth2 token.
     * @return instance of OAuth1Token setter.
     */
    public static OAuth1Token token(String token) {
        return new OAuth1Token(token);
    }
}

File: src/main/kotlin/com/github/aivancioglo/resttest/http/OAuth2.kt
package com.github.aivancioglo.resttest.http

import com.github.aivancioglo.resttest.http.OAuthSignature.HEADER
import com.github.aivancioglo.resttest.setters.Setter

/**
 * Class for OAuth 2 using in requests.
 */
class OAuth2 {
    var used = false

    var token: String = ""
        set(value) {
            field = value
            used = true
        }

    var signature: OAuthSignature = HEADER
        set(value) {
            field = value
            used = true
        }

    companion object {

        /**
         * Setter of OAuth 2 token.
         *
         * @return consumer key.
         */
        @JvmStatic
        fun token(token: String) = object : Setter {
            override fun update(request: Request) {
                request.oAuth2.token = token
            }
        }

        /**
         * Setter of OAuth 2 token signature
         *
         * @return consumer key.
         */
        @JvmStatic
        fun signature(signature: OAuthSignature) = object : Setter {
            override fun update(request: Request) {
                request.oAuth2.signature = signature
            }
        }
    }
}
